internal/runners/activate/activate.go

```diff
@@ -90,33 +110,29 @@ func (r *Activate) run(params *ActivateParams, activatorLoop activationLoopFunc)
 		r.subshell.SetActivateCommand(params.Command)
 	}
 
-	return activatorLoop(r.out, r.subshell, targetPath, activate)
+	return activatorLoop(r.out, r.subshell, projectPath, activate)
 }
 
-func (r *Activate) setupPath(namespace string, preferredPath string) (string, error) {
-	var (
-		proj *project.Project
-		fail *failures.Failure
-	)
-
+func (r *Activate) pathToUse(namespace string, preferredPath string) (string, error) {
 	switch {
-	// Checkout via namespace (eg. state activate org/project) and set resulting path
 	case namespace != "":
-		namesPath, err := r.namespaceSelect.Run(namespace, preferredPath)
-		if err != nil {
-			return "", err
-		}
-		proj, fail = project.FromPath(namesPath)
-	// Use the user provided path
+		// Checkout via namespace (eg. state activate org/project) and set resulting path
+		return r.namespaceSelect.Run(namespace, preferredPath)
 	case preferredPath != "":
-		proj, fail = project.FromPath(preferredPath)
-	// normal project getter - falls back to default project if none found in working directory
+		// Use the user provided path
+		return preferredPath, nil
 	default:
-		proj, fail = project.GetSafe()
-	}
-	if fail != nil {
-		return "", fail.ToError()
+		// Get path from working directory
+		var fail *failures.Failure
+		targetPath, fail := projectfile.GetProjectFilePath()
+		return filepath.Dir(targetPath), fail.ToError()
```

Comment:
An error return along with useful value is quite abnormal. This would usually be solved by returning the useful value as part of a specific error return so that callers know they are in a special condition. Barring that for whatever reason, I suggest including a comment on the function clarifying the unusual behavior.

If the return here is misleading me, I suggest explicitly returning empty string w/error or processed string with nil.

Reply:
This function has no opinion on whether the value is valid, it's just returning the error from `GetProjectFilePath()`. The fail.ToError() stuff is just obscuring things, but that will go away when we get rid of failures.

I don't think this needs to be any more verbose than it already is. Once failures go away this'll be nice and neat, until then it seems like a pretty minor stumbling block.

Reply:
My only significant concern here is clarifying if callers are to be interacting with the useful value when there is also an error return. It's clean and readable enough otherwise, IMO.

Reply:
Error always takes precedence. Callers don't need any special logic.